pkg/util/policy/agent/agent.py

```diff
@@ -284,138 +258,231 @@ def make_new_full_ruleset(current_rules, new_rules):
+            # consider policy to be tenant wide.
+            target_segment_forward_chain = tenant_wide_policy_vector_chain
+
+        # Chain names are going to be used later to fill in the rules. Store them.
+        policy_chains[in_chain_name] = True
+
+        # Jump from per-tenant chain into per-segment chains and default DROP.
+        rules[tenant_policy_vector_chain] = [
+            _make_rule(tenant_policy_vector_chain, "-j %s" % target_segment_forward_chain),
+            _make_rule(tenant_policy_vector_chain, "-j %s" % tenant_wide_policy_vector_chain),
+            _make_rule(tenant_policy_vector_chain, "-j DROP")
+        ]
+
+        # Jump from per-segment chain into policy chain
+        if target_segment_forward_chain not in rules:
+            rules[target_segment_forward_chain] = []
+            rules[target_segment_forward_chain].append(
+                    _make_rule(target_segment_forward_chain, '-m comment --comment POLICY_CHAIN_HEADER -j RETURN'))
+        rules[target_segment_forward_chain].insert(0,
+                _make_rule(target_segment_forward_chain, "-j %s" % policy_chain_name))
+
+        # Default rules per tenant.
+        if tenant_wide_policy_vector_chain not in rules:
+            rules[tenant_wide_policy_vector_chain] = []
+            rules[tenant_wide_policy_vector_chain].append(
+                _make_rule(tenant_wide_policy_vector_chain, '-m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT'))
+            rules[tenant_wide_policy_vector_chain].append(
+                _make_rule(tenant_wide_policy_vector_chain, '-m comment --comment POLICY_CHAIN_HEADER -j RETURN'))
+
+        # Loop over peers and fill top level policy chains with source matching rules
+        for peer in policy_def.get('peers'):
+            # Possible peers are
+            # CIDR - detected by cidr field
+            # peer - detected by peer type
+            # tid/sid - detected by tenant_network_id and segment_network_id
+            pr = peer.get('peer')
+            cidr = peer.get('cidr')
+            from_tenant = peer.get('tenant_network_id')
+            from_segment = peer.get('segment_network_id')
+
+            if pr:
+                if pr == "any":
+                    jump_rules = [ _make_rule(policy_chain_name, "-j %s") % in_chain_name ]
+                else:
+                    raise Exception("Unsupported value of peer %s" % pr)
+
+            elif cidr:
+                jump_rules = [ _make_rule(policy_chain_name, "-s %s -j %s") % (cidr, in_chain_name) ]
+
+            elif from_segment and from_tenant:
+                u32_in_match = _make_u32_match(addr_scheme, tenant, from_segment)
+                jump_rules = [
+                    _make_rule(policy_chain_name, '-m u32 --u32 "%s" -j %s' %
+                                                       (u32_in_match, in_chain_name))
+                ]
+
+            else:
+                raise Exception("Unknown peer type %s" % pr)
```

Comment:
I guess it would be useful.

But how could it be 'None'? You're only in that loop if it's something?

It would be nice to have extra handling for the case where tenant is defined, but segment is not (or the other way around), since currently you might get an error telling you that 'from_tenant' is an unknown peer type, which will seem strange to the user.